### DISCUSSION TITLE
generateUserImports: Include extensions when importing fixtures

### DISCUSSION THREAD

**Discussion by mwilliamson** [react-cosmos/react-cosmos#1661] [Ideas]
Hello!

I have a TypeScript project with "module" set to "node16" (since the package includes code both for the browser and for node.js.) The file generated by the `--expose-imports` flag strips the `.tsx` extension from the imported fixtures, but this causes an error in TypeScript since the extension is needed for any imports.

I think I've fixed this locally by changing the extension replacement in `userImportsTemplate` and `userImportsLazyTemplate` from:

```
return typeScript ? filePath.replace(/\.tsx?$/, '') : filePath;
```

to:

```
return typeScript ? filePath.replace(/\.tsx?$/, '.js') : filePath;
```

I'm definitely not a JavaScript modules expert, but I think this is generally safe to do given other package types allow extensions even if they don't require it? Alternatively, would it make sense to add an option for this?

Thanks!

**Comment by ovidiuch** [comment]
Hi @mwilliamson,

If only I could remember why I made this change 19 months ago: https://github.com/react-cosmos/react-cosmos/pull/1466 :)

The only comment that I left in that PR doesn't make a lot of sense to me right now:

> Add compat with projects where allowImportingTsExtensions isn't enabled

Not sure how [allowImportingTsExtensions](https://www.typescriptlang.org/tsconfig/#allowImportingTsExtensions) relates to .js extensions. The comment makes more sense for the reason why the original ts/tsx extension can't be kept, though.

On the other hand, I just tried make the change you propose (essentially reverting #1466) in an Expo project and Metro doesn't seem to be able to resolve .js extensions for TS modules. So the safe way to go is probably to make this behavior configurable.
